src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,304 @@
+    uint16_t                       startOffset  = aMessageOut.GetLength();
+
+    VerifyOrExit(aLinkMetricsQuery != nullptr && aLinkMetricsQuery->GetLength() > 0 && aLinkMetricsQuery->IsValid(),
+                 error = OT_ERROR_PARSE);
+    queryId = aLinkMetricsQuery->GetQueryId();
+    VerifyOrExit(queryId->IsValid(), error = OT_ERROR_PARSE);
+    VerifyOrExit(queryId->GetLength() > 0 && queryId->GetSeriesId() < 255, error = OT_ERROR_INVALID_ARGS);
+
+    // Link Metrics Report TLV
+    tlv.SetType(Mle::Tlv::kLinkMetricsReport);
+    SuccessOrExit(error = aMessageOut.Append(&tlv, sizeof(tlv)));
+
+    if (queryId->GetSeriesId() == 0)
+    {
+        queryOptions = aLinkMetricsQuery->GetQueryOptions();
+        VerifyOrExit(queryOptions->IsValid(), error = OT_ERROR_PARSE);
+        SuccessOrExit(error = AppendSingleProbeLinkMetricsReport(aMessageIn, queryOptions,
+                                                                 Get<Mac::Mac>().GetNoiseFloor(), aMessageOut, length));
+    }
+    else
+    {
+        ExitNow(error = OT_ERROR_NOT_IMPLEMENTED);
+    }
+
+    tlv.SetLength(length);
+    aMessageOut.Write(startOffset, sizeof(tlv), &tlv);
+
+exit:
+    return error;
+}
+
+void LinkMetrics::HandleLinkMetricsReport(const Message &     aMessage,
+                                          uint16_t            aOffset,
+                                          uint16_t            aLength,
+                                          const Ip6::Address &aAddress)
+{
+    otLinkMetric      metrics[kLinkMetricsMaxTypeIdFlags];
+    uint8_t           metricsCount = 0;
+    uint16_t          pos          = aOffset;
+    uint16_t          end_pos      = aOffset + aLength;
+    Tlv               tlv;
+    LinkMetricsTypeId typeId;
+
+    aMessage.Read(pos, sizeof(Tlv), &tlv);
+    VerifyOrExit(tlv.GetType() == kLinkMetricsReportSub, OT_NOOP);
+
+    while (pos < end_pos)
+    {
+        aMessage.Read(pos, sizeof(Tlv), &tlv);
+        pos += sizeof(Tlv);
+        aMessage.Read(pos, sizeof(LinkMetricsTypeId), &typeId);
+        pos += sizeof(otLinkMetricsTypeId);
+        SetLinkMetricsTypeIdFromTlv(metrics[metricsCount].mLinkMetricsTypeId, typeId);
+
+        if (metrics[metricsCount].mLinkMetricsTypeId.mLinkMetricsFlagL)
+        {
+            aMessage.Read(pos, sizeof(uint32_t), &metrics[metricsCount].mLinkMetricValue.m32);
+            pos += sizeof(uint32_t);
+        }
+        else
+        {
+            aMessage.Read(pos, sizeof(uint8_t), &metrics[metricsCount].mLinkMetricValue.m8);
+            pos += sizeof(uint8_t);
+        }
+
+        metricsCount++;
+    }
+
+    if (mLinkMetricsReportCallback != nullptr)
+    {
+        mLinkMetricsReportCallback(&aAddress, metrics, metricsCount, mContext);
+    }
+
+exit:
+    return;
+}
+
+void LinkMetrics::SetLinkMetricsReportCallback(otLinkMetricsReportCallback aCallback, void *aCallbackContext)
+{
+    mLinkMetricsReportCallback = aCallback;
+    mContext                   = aCallbackContext;
```

Comment:
```suggestion
    mLinkMetricsReportCallbackContext                   = aCallbackContext;
```

nit: Rename to be more explicit ?